### ISSUE TITLE
segment violation on Grace Hopper aarch64

### ISSUE THREAD

**Issue by lunitin** [plazonic/nvidia_gpu_prometheus_exporter#1]
Execution issues on a GraceHopper running Rocky 9.4 with the aarch64+64k kernel...

```
root@gh01 nvidia_gpu_prometheus_exporter]# nvidia-smi
Tue Aug 27 15:33:05 2024
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.35.03              Driver Version: 560.35.03      CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GH200 480GB             On  |   00000009:01:00.0 Off |                    0 |
| N/A   26C    P0             68W /  900W |       1MiB /  97871MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+

[root@gh01 nvidia_gpu_prometheus_exporter]# uname -r
5.14.0-427.31.1.el9_4.aarch64+64k

[root@gh01 nvidia_gpu_prometheus_exporter]# go build

[root@gh01 nvidia_gpu_prometheus_exporter]# ./nvidia_gpu_prometheus_exporter
SIGSEGV: segmentation violation
PC=0x0 m=0 sigcode=1
signal arrived during cgo execution

goroutine 1 [syscall]:
runtime.cgocall(0x7a4150, 0x4000173e18)
	/usr/lib/golang/src/runtime/cgocall.go:157 +0x44 fp=0x4000173de0 sp=0x4000173da0 pc=0x407e94
github.com/NVIDIA/go-nvml/pkg/nvml._Cfunc_nvmlInit_v2()
	_cgo_gotypes.go:4597 +0x34 fp=0x4000173e10 sp=0x4000173de0 pc=0x696c04
github.com/NVIDIA/go-nvml/pkg/nvml.nvmlInit_v2()
	/root/go/pkg/mod/github.com/!n!v!i!d!i!a/go-nvml@v0.12.0-0/pkg/nvml/nvml.go:32 +0x1c fp=0x4000173e30 sp=0x4000173e10 pc=0x6975cc
github.com/NVIDIA/go-nvml/pkg/nvml.Init()
	/root/go/pkg/mod/github.com/!n!v!i!d!i!a/go-nvml@v0.12.0-0/pkg/nvml/init.go:43 +0x9c fp=0x4000173e60 sp=0x4000173e30 pc=0x696e8c
main.main()
	/root/nvidia_gpu_prometheus_exporter/main.go:357 +0x70 fp=0x4000173f30 sp=0x4000173e60 pc=0x79ea10
runtime.main()
	/usr/lib/golang/src/runtime/proc.go:267 +0x2bc fp=0x4000173fd0 sp=0x4000173f30 pc=0x43dbec
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_arm64.s:1197 +0x4 fp=0x4000173fd0 sp=0x4000173fd0 pc=0x46e514

goroutine 2 [force gc (idle)]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/usr/lib/golang/src/runtime/proc.go:398 +0xc8 fp=0x40000eaf90 sp=0x40000eaf70 pc=0x43e018
runtime.goparkunlock(...)
	/usr/lib/golang/src/runtime/proc.go:404
runtime.forcegchelper()
	/usr/lib/golang/src/runtime/proc.go:322 +0xb8 fp=0x40000eafd0 sp=0x40000eaf90 pc=0x43dea8
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_arm64.s:1197 +0x4 fp=0x40000eafd0 sp=0x40000eafd0 pc=0x46e514
created by runtime.init.6 in goroutine 1
	/usr/lib/golang/src/runtime/proc.go:310 +0x24

goroutine 3 [GC sweep wait]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/usr/lib/golang/src/runtime/proc.go:398 +0xc8 fp=0x40000eb760 sp=0x40000eb740 pc=0x43e018
runtime.goparkunlock(...)
	/usr/lib/golang/src/runtime/proc.go:404
runtime.bgsweep(0x0?)
	/usr/lib/golang/src/runtime/mgcsweep.go:280 +0xa0 fp=0x40000eb7b0 sp=0x40000eb760 pc=0x428f20
runtime.gcenable.func1()
	/usr/lib/golang/src/runtime/mgc.go:200 +0x28 fp=0x40000eb7d0 sp=0x40000eb7b0 pc=0x41da38
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_arm64.s:1197 +0x4 fp=0x40000eb7d0 sp=0x40000eb7d0 pc=0x46e514
created by runtime.gcenable in goroutine 1
	/usr/lib/golang/src/runtime/mgc.go:200 +0x6c

goroutine 4 [GC scavenge wait]:
runtime.gopark(0x400002e070?, 0x934ac0?, 0x1?, 0x0?, 0x40000031e0?)
	/usr/lib/golang/src/runtime/proc.go:398 +0xc8 fp=0x40000ebf50 sp=0x40000ebf30 pc=0x43e018
runtime.goparkunlock(...)
	/usr/lib/golang/src/runtime/proc.go:404
runtime.(*scavengerState).park(0xc69ac0)
	/usr/lib/golang/src/runtime/mgcscavenge.go:425 +0x5c fp=0x40000ebf80 sp=0x40000ebf50 pc=0x4267ec
runtime.bgscavenge(0x0?)
	/usr/lib/golang/src/runtime/mgcscavenge.go:653 +0x44 fp=0x40000ebfb0 sp=0x40000ebf80 pc=0x426d34
runtime.gcenable.func2()
	/usr/lib/golang/src/runtime/mgc.go:201 +0x28 fp=0x40000ebfd0 sp=0x40000ebfb0 pc=0x41d9d8
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_arm64.s:1197 +0x4 fp=0x40000ebfd0 sp=0x40000ebfd0 pc=0x46e514
created by runtime.gcenable in goroutine 1
	/usr/lib/golang/src/runtime/mgc.go:201 +0xac

goroutine 5 [finalizer wait]:
runtime.gopark(0x40001824e0?, 0x1a0?, 0x98?, 0x1?, 0x884d00?)
	/usr/lib/golang/src/runtime/proc.go:398 +0xc8 fp=0x40000ea580 sp=0x40000ea560 pc=0x43e018
runtime.runfinq()
	/usr/lib/golang/src/runtime/mfinal.go:193 +0x108 fp=0x40000ea7d0 sp=0x40000ea580 pc=0x41ca68
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_arm64.s:1197 +0x4 fp=0x40000ea7d0 sp=0x40000ea7d0 pc=0x46e514
created by runtime.createfing in goroutine 1
	/usr/lib/golang/src/runtime/mfinal.go:163 +0x80

r0      0x4000174000
r1      0x7a4150
r2      0x4000173d90
r3      0xc69de0
r4      0x270
r5      0x4000002230
r6      0x1
r7      0x65636e6174736e49
r8      0xc6a240
r9      0x4000173e18
r10     0xffff0000afb1f937
r11     0x7
r12     0x736e6563694c7465
r13     0xffffc7a7d960
r14     0x0
r15     0x0
r16     0x40001723a0
r17     0xffffaa0a0f20
r18     0x0
r19     0x4000173e18
r20     0x4000173d00
r21     0x4000174000
r22     0x1
r23     0x8e41dc0be9fcd734
r24     0xf060d7b75027bdc5
r25     0x7973ad2e0f
r26     0x8c33e8
r27     0x10
r28     0xc69de0
r29     0xffffc7a7d930
lr      0x7a4170
sp      0xffffc7a7d930
pc      0x0
fault   0x0
```

Any suggestions?

**Comment by ggaffield** [comment]
I ran into the same problem, only on x86_64 with 5.14.0-427.33.1.el9_4.x86_64.
See the discussion here: https://github.com/golang/go/issues/63264

TL;DR, you need a more recent version of go-nvml.
This can be done by doing: 
```
go get -u
go mod tidy
go build
```

**Comment by l1ll1** [comment]
Came here to say this, found the one open issue already addressed my problem and I should have check the issues before googling the segv ... Having said that I used
```
go get github.com/NVIDIA/go-nvml@v0.12.0-7
```
rather than updating everything

@plazonic do you think you could do a merge just to bump the version on go-nvml pretty please; just to save other some time ;-)

**Comment by plazonic** [comment]
I updated all deps as suggested by [ggaffield](https://github.com/ggaffield). Thanks to both of you!
